lib/cli/src/Cardano/CLI.hs

```diff
@@ -130,14 +212,371 @@ import System.IO
     , stdout
     , utf8
     )
+import Text.Heredoc
+    ( here )
 
 import qualified Data.Aeson as Aeson
+import qualified Data.Aeson.Encode.Pretty as Aeson
 import qualified Data.Aeson.Types as Aeson
 import qualified Data.ByteString.Lazy as BL
+import qualified Data.ByteString.Lazy.Char8 as BL8
 import qualified Data.List.NonEmpty as NE
 import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
 import qualified Data.Text.IO as TIO
 
+{-------------------------------------------------------------------------------
+                                Environment
+-------------------------------------------------------------------------------}
+
+-- | A convenient representation of a complete CLI environment.
+data Environment = Environment
+    { args
+        :: Arguments
+    , cli
+        :: Docopt
+    , argPresent
+        :: Option -> Bool
+    , parseAllArgs
+        :: forall a . FromText a => Option -> IO (NE.NonEmpty a)
+    , parseArg
+        :: forall a . FromText a => Option -> IO a
+    , parseOptionalArg
+        :: forall a . FromText a => Option -> IO (Maybe a)
+    }
+
+{-------------------------------------------------------------------------------
+                              CLI Construction
+-------------------------------------------------------------------------------}
+
+-- | Make a specialized CLI from the specified commands, options, and examples.
+--
+makeCli :: String -> String -> String -> Docopt
+makeCli cliCommandsSpecific cliOptionsSpecific cliExamplesSpecific =
+    fromRight (error "Unable to construct CLI.") $
+        parseUsage usageString
+  where
+    usageString = mempty
+        <> cliHeader
+        <> "\nUsage:\n"
+        <> cliCommands
+        <> "\nOptions:\n"
+        <> cliOptions
+        <> "\nExamples:\n"
+        <> cliExamples
+    cliCommands = unlines $ filter (not . null) $ lines $ mempty
+        <> cliCommandsSpecific
+        <> cliCommandsGeneric
+    cliOptions = unlines $ filter (not . null) $ sort $ lines $ mempty
+        <> cliOptionsSpecific
+        <> cliOptionsGeneric
+    cliExamples = unlines $ filter (not . null) $ lines $ mempty
+        <> cliExamplesSpecific
+        <> cliExamplesGeneric
```

Comment:
:thinking: I think we're loosing the benefits of using `docopt` in a first place. I understand the need for removing duplication between both files yet I am not sure we are taking it to the right direction :/ .. It's probably the right time to switch to `optparse-applicative` for the CLI.

Reply:
>  I think we're loosing the benefits of using docopt in a first place. I understand the need for removing duplication between both files yet I am not sure we are taking it to the right direction :/ .. It's probably the right time to switch to optparse-applicative for the CLI.

I completely agree. Looking forward to trying a different direction with `optparse-applicative` here.